core/src/main/scala/cats/ApplicativeError.scala

```diff
@@ -287,6 +287,26 @@ trait ApplicativeError[F[_], E] extends Applicative[F] {
   def catchOnly[T >: Null <: Throwable]: CatchOnlyPartiallyApplied[T, F, E] =
     new CatchOnlyPartiallyApplied[T, F, E](this)
 
+  /**
+   * Often E can be created from Throwable. Here we try to call pure or
+   * catch, adapt into E, and raise.
+   *
+   * Exceptions that cannot be adapted to E will be propagated
+   */
+  def catchNonFatalAs[A](adaptIfPossible: Throwable => Option[E])(a: => A): F[A] =
+    try pure(a)
+    catch {
+      case e if NonFatal(e) => adaptIfPossible(e).map(raiseError[A]).getOrElse(throw e)
```

Comment:
This line concerns me quite a lot, particularly `throw e`.
It means that the exception will be re-thrown out of the `F` context, which is not what I would expect from any `ApplicativeError` instance.

Reply:
Unless we require the adaptation function to be total, we're going to need to rethrow these, because there isn't a way to raise a `Throwable` into an arbitrary `ApplicativeError[F, E]`

Reply:
Yes, exactly. This is why we must not use a `PartialFunction` as you suggested in https://github.com/typelevel/cats/pull/4616/files#r1633942586. See my comments in https://github.com/typelevel/cats/issues/4286#issuecomment-1226096402.

Reply:
@armanbilge , `PartialFunction[A, B]` and `A => Option[B]` are isomorphic. Do I understand correctly your point that we must not use either of them?